Logic for collecting Histogram efficiently using Point Trees

jainankitk opened the pull request:
### Description

This PR adds multi range traversal logic to collect the histogram on numeric field indexed as pointValues for MATCH_ALL cases. Even for non-match all cases like `PointRangeQuery`, if the query field == histogram field, this logic can be used. For the later, need to supply the `PointRangeQuery` bounds for building the appropriate `Ranges` to be collected. Need some inputs from the community on how it can be plugged correctly into the `HistogramCollector`

One of the key assumptions is absence of any deleted documents. Maybe going forward (especially if the deleted documents percentage is low), we can consider correcting the collected `Ranges` by subtracting for deleted documents. Although if I remember correctly, getting doc values for just deleted documents was non-trivial task!

Related issue https://github.com/apache/lucene/issues/13335

jainankitk committed:
Adding logic for collecting histogram using Point Trees

jainankitk committed:
Adding license and NonFinalStaticField fix

jainankitk committed:
Fixing static field error

jainankitk committed:
Tidying the code

jainankitk commented:
@stefanvodita / @jpountz - Would love to get your thoughts on this optimization, and how we can leverage it in Lucene. In a nutshell, it solves the below problem:

Given a sorted non-overlapping set of intervals (Histogram buckets could be an example), it collects the matching documents count in single travel of `PointsTree` index, by skipping over the `leafBlocks` completely unless the values in `leafBlock` overlap with more than one interval. This ensures that the `# leafBlocks` actually traversed is bounded by the `# buckets` and remaining `leafBlocks` are collected in bulk. Hence it can very efficiently collect the doc counts, especially when the `# documents / # buckets` is pretty high.

stefanvodita reviewed (commented):
Sorry, I only had a quick look. Is the opimisation here analogous to the one `HistogramLeafCollector` does with the skipper?

jainankitk commented:
> Sorry, I only had a quick look. Is the opimisation here analogous to the one `HistogramLeafCollector` does with the skipper?

No, this approach is different from the skipper as it leverages `PointValues` instead of `DocValues` for computing the buckets

stefanvodita commented:
I didn't mean to imply that the two solutions are the same, apologies if that's how it came across.

> Need some inputs from the community on how it can be plugged correctly into the HistogramCollector

Let me know if this doesn't answer the question @jainankitk, maybe you'd already gone through this and you were looking for a different answer.
I think you could start in `HistogramCollector.getLeafCollector` ([code](https://github.com/apache/lucene/blob/4957766fcee52c534d786e3948fadf6d36c9779f/lucene/sandbox/src/java/org/apache/lucene/sandbox/facet/plain/histograms/HistogramCollector.java#L50)). Right now we throw an exception if the field we're using isn't doc values ([code](https://github.com/apache/lucene/blob/4957766fcee52c534d786e3948fadf6d36c9779f/lucene/sandbox/src/java/org/apache/lucene/sandbox/facet/plain/histograms/HistogramCollector.java#L59)). You'd need a new branch for the case you want to implement and a new `LeafCollector`, similar to the ones already in the file. Having that would make it easier to think through the next steps.

At a higher level, I'm curious if you had a use-case in mind.

jainankitk commented:
> I didn't mean to imply that the two solutions are the same, apologies if that's how it came across.

Not at all. Even I was initially confused with skipper logic, only after spending some time realized this approach is slightly different. So, thanks for reiterating the question.

> I think you could start in HistogramCollector.getLeafCollector ([code](https://github.com/apache/lucene/blob/4957766fcee52c534d786e3948fadf6d36c9779f/lucene/sandbox/src/java/org/apache/lucene/sandbox/facet/plain/histograms/HistogramCollector.java#L50)). Right now we throw an exception if the field we're using isn't doc values ([code](https://github.com/apache/lucene/blob/4957766fcee52c534d786e3948fadf6d36c9779f/lucene/sandbox/src/java/org/apache/lucene/sandbox/facet/plain/histograms/HistogramCollector.java#L59)).

Currently, `Collector` doesn't need to be aware of the `Query` itself. They are designed to collect individual docId or using `DocIdStream` from the scorer. But this `CustomCollector`, does not need the scorer to provide documents, but can `BulkCollect` documents, assuming `MATCH_ALL` or `PointRangeQuery` (where `PointRangeQuery.field == histogram.field`). Otherwise, it should fallback to traditional methods for collecting matching documents.


> At a higher level, I'm curious if you had a use-case in mind.

This optimization can be applied to following use cases:
* Number of sale based on the price range (0-50, 50-100, 100-250,.....)
* Number of visits on website for each day in a month

Just as a data point, this change helped us improve date histogram latency from 5168 ms to 160 ms (~32x!!) for big5 workload in OpenSearch

jainankitk committed:
Enabling PointTreeBulkCollector using query in weight and adding unit test

Signed-off-by: Ankit Jain <jain.ankitk@gmail.com>

jainankitk committed:
Removing usage of java random

Signed-off-by: Ankit Jain <jain.ankitk@gmail.com>

jainankitk commented:
I have updated the PR, and the code flow is like below now:

* `HistogramCollector` overrides the `setWeight` for accessing the underlying `Query`
* To keep things simple, just optimizing for `MATCH_ALL_DOCS` query and no deleted documents for now
* Optimized path is enabled only if `pointTree` is built for the field
* There are few other conditions for optimized path being enabled. Being conservative for now, and fallback to original.
* Add small unit test to verify working as expected

Will add few more unit tests, once I get some feedback on the code changes. Can also include small performance unit test that demonstrates, pointTree based collection is faster than the docValues based collector.

stefanvodita commented on lucene/sandbox/src/java/org/apache/lucene/sandbox/facet/plain/histograms/PointTreeBulkCollector.java:
```diff
@@ -0,0 +1,219 @@
+    // We want that # leaf nodes is more than # buckets so that we can completely skip over
+    // some of the leaf nodes. Higher this ratio, more efficient it is than naive approach!
+    if ((pointValues.size() / 512) < (leafMaxBucket - leafMinBucket)) {
+      return false;
```
Similar question as above. What will the user get in this case?

jainankitk replied:
The user does not need to get anything. We will just skip the optimized path and fall back to the original execution plan

stefanvodita replied:
I could be misunderstanding, so please bear with me. Let's say a user only has `PointValues` indexed and no `DocValues` and they try to use the histogram. Would they be able to do that even if their data doesn't fit the efficiency condition above?

jainankitk replied:
Ah, not really. As explained in other, we can discuss that as part of https://github.com/apache/lucene/issues/14536?

stefanvodita reviewed (commented):
Thanks for iterating on this @jainankitk!

jainankitk commented on lucene/sandbox/src/java/org/apache/lucene/sandbox/facet/plain/histograms/PointTreeBulkCollector.java:
```diff
@@ -0,0 +1,219 @@
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.util.ArrayUtil;
+
+class PointTreeBulkCollector {
```
Do we need explicit `@lucene.experimental` even for classes within sandbox? As per my understanding, everything sandbox is experimental by default?

stefanvodita replied:
There are other files in sandbox with experimental tags, so maybe better safe then sorry? I guess it's easy to forget to add those when moving a piece of code out of sandbox.

jainankitk replied:
> There are other files in sandbox with experimental tags, so maybe better safe then sorry?

Yup, adding the tag!

jainankitk commented on lucene/sandbox/src/java/org/apache/lucene/sandbox/facet/plain/histograms/PointTreeBulkCollector.java:
```diff
@@ -0,0 +1,219 @@
+        || pointValues.getNumDimensions() != 1
+        || pointValues.getDocCount() != pointValues.size()
+        || ArrayUtil.getValue(pointValues.getBytesPerDimension()) == null) {
+      return false;
```
The user does not need to get anything. We will just skip the optimized path and fall back to the original execution plan

stefanvodita replied:
Let me be more specific. I'm questioning if the `Expected numeric field, but got doc-value type` exception we used to throw still makes sense now or needs to be adjusted since we now have another case for `PointValues` which must have not worked out to reach the exception.

jainankitk replied:
I think we can keep this for now as is. The way to read it is, it works same as before including exceptions and everything. There is possibly optimized path if pointValues are indexed. I have created another issue for tackling this specific scenario - https://github.com/apache/lucene/issues/14536. Might need some more changes or slightly different approach to `PointTreeBulkCollector`

stefanvodita replied:
I see your point and I'm happy to have a follow-up issue if necessary for this and the related comment below.

jainankitk commented on lucene/sandbox/src/test/org/apache/lucene/sandbox/facet/plain/histograms/TestHistogramCollectorManager.java:
```diff
@@ -115,6 +117,38 @@ public void testSkipIndexWithSortAndLowInterval() throws IOException {
             .setCodec(TestUtil.alwaysDocValuesFormat(new Lucene90DocValuesFormat(3))));
   }
 
+  public void testMultiRangePointTreeCollector() throws IOException {
```
Okay, ran the test and few times and I understand what the error is.

```
java.lang.AssertionError: expected:<[1=>978, 3=>1004, 4=>1000, 2=>1022, 0=>996]> but was:<[4=>5000, 1=>4890, 2=>5110, 3=>5020, 0=>4980]>
```

The actual values are 5 times of the expected. Essentially, the logic efficiently collects values for entire segment and in case of intra-segment concurrency, it ends up collecting 5 times. I am wondering if the collectors can be aware of intra-segment concurrency, and not use the optimized path when it is enabled?

jainankitk replied:
Okay, I have added logic for collecting only for one of the partitions if intra-segment concurrency is enabled. Ran the test 50k times after that without any failures:

```
> Task :lucene:core:testClasses
> Task :lucene:core:test
> Task :lucene:core:wipeTaskTemp

> Task :lucene:sandbox:test
:lucene:sandbox:test (SUCCESS): 50000 test(s)

> Task :lucene:sandbox:wipeTaskTemp
The slowest suites (exceeding 1s) during this run:
  580.41s TestHistogramCollectorManager (:lucene:sandbox)

BUILD SUCCESSFUL in 9m 58s
246 actionable tasks: 181 executed, 65 up-to-date
```

jainankitk committed:
Addressing review comments

Signed-off-by: Ankit Jain <jain.ankitk@gmail.com>

jainankitk committed:
Handling intra segment better using concurrent map instead of static synchronized

Signed-off-by: Ankit Jain <jain.ankitk@gmail.com>

jainankitk commented on lucene/core/src/java/org/apache/lucene/util/ArrayUtil.java:
```diff
@@ -801,4 +802,16 @@ public static int compareUnsigned4(byte[] a, int aOffset, byte[] b, int bOffset)
         (int) BitUtil.VH_BE_INT.get(a, aOffset), (int) BitUtil.VH_BE_INT.get(b, bOffset));
   }
+
+  public static Function<byte[], Long> getValue(int numBytes) {
```
Thanks for suggesting `NumericUtils`. I wasn't aware of the `NumericUtils#sortableBytesToLong`, exactly what I needed to flip the sign bit correctly!

```
// Flip the sign bit back
    v ^= 0x8000000000000000L;
```

I have removed this method from `ArrayUtil` and added in `PointTreeBulkCollector`. Please let me know if you still see any concerns

stefanvodita replied:
I'm happy with how it is now. Thank you!

jainankitk commented on lucene/sandbox/src/java/org/apache/lucene/sandbox/facet/plain/histograms/HistogramCollector.java:
```diff
@@ -53,11 +59,22 @@ public LeafCollector getLeafCollector(LeafReaderContext context) throws IOExcept
+    // We can use multi range traversal logic to collect the histogram on numeric
+    // field indexed as point range query for MATCH_ALL cases
+    final PointValues pointValues = context.reader().getPointValues(field);
+    if (query instanceof MatchAllDocsQuery && !context.reader().hasDeletions()) {
```
Yes, I will open follow-up once we are close to merging this PR. I will keep this conversation open, until the issue is created

jainankitk committed:
Tidying the code

jainankitk committed:
Running small performance test on existing vs new approach

Signed-off-by: Ankit Jain <jain.ankitk@gmail.com>

jainankitk commented:
Addressed most of the review comments. Ran small performance benchmark to see the difference, and we can see some even with small number of documents:

Bulk time is new approach, and skip index time is the existing
With 100k docs, and 5k bucketWidth
```
  1> Bulk time was : 10593417 vs skip index time : 12941417
  1> Bulk time was : 1006750 vs skip index time : 3241959
  1> Bulk time was : 770916 vs skip index time : 4690875
  1> Bulk time was : 870917 vs skip index time : 2621291
  1> Bulk time was : 754709 vs skip index time : 3175500
  1> Bulk time was : 1389750 vs skip index time : 2557125
  1> Bulk time was : 788125 vs skip index time : 4633541
  1> Bulk time was : 767125 vs skip index time : 5078917
  1> Bulk time was : 727208 vs skip index time : 6333083
  1> Bulk time was : 755250 vs skip index time : 3179917
```

With 250k documents, and 10k bucketWidth:
```
  1> Bulk time was : 14125250 vs skip index time : 20235666
  1> Bulk time was : 915583 vs skip index time : 15929834
  1> Bulk time was : 886959 vs skip index time : 15842375
  1> Bulk time was : 904667 vs skip index time : 10911500
  1> Bulk time was : 801959 vs skip index time : 10846750
  1> Bulk time was : 844834 vs skip index time : 14980917
  1> Bulk time was : 855209 vs skip index time : 5865500
  1> Bulk time was : 804334 vs skip index time : 15966667
  1> Bulk time was : 805042 vs skip index time : 11613791
  1> Bulk time was : 830958 vs skip index time : 11036833
```

As we can see the difference is bit more pronounced with increasing number of documents and wide enough bucketWidth.

jainankitk committed:
Adding jmh benchmark for histogram collection

Signed-off-by: Ankit Jain <jain.ankitk@gmail.com>

jainankitk committed:
Fixing forbidden class error

Signed-off-by: Ankit Jain <jain.ankitk@gmail.com>

jainankitk committed:
Fixing unused parameter error

Signed-off-by: Ankit Jain <jain.ankitk@gmail.com>

jainankitk commented:
@stefanvodita - Thanks for a prompt review. Addressed most of the review comments. Adding JMH benchmark instead of the not so useful performance test added earlier. The benchark results demonstrate significant increase in throughput with increasing # documents and bucket width (lesser buckets mean less low level traversal in point range tree and more documents collected in bulk):

```
Benchmark                                     (bucketWidth)  (docCount)  (pointEnabled)   Mode  Cnt     Score      Error  Units
HistogramCollectorBenchmark.collectHistogram           5000      500000            true  thrpt    3  2114.002 ±   46.852  ops/s
HistogramCollectorBenchmark.collectHistogram           5000      500000           false  thrpt    3  1236.532 ±  724.312  ops/s
HistogramCollectorBenchmark.collectHistogram           5000     5000000            true  thrpt    3   220.636 ±   17.688  ops/s
HistogramCollectorBenchmark.collectHistogram           5000     5000000           false  thrpt    3   147.054 ±   37.055  ops/s
HistogramCollectorBenchmark.collectHistogram          25000      500000            true  thrpt    3  9788.181 ± 4388.445  ops/s
HistogramCollectorBenchmark.collectHistogram          25000      500000           false  thrpt    3  1227.537 ±   21.222  ops/s
HistogramCollectorBenchmark.collectHistogram          25000     5000000            true  thrpt    3   856.278 ±  840.625  ops/s
HistogramCollectorBenchmark.collectHistogram          25000     5000000           false  thrpt    3   145.322 ±   86.603  ops/s
```

jpountz commented on lucene/sandbox/src/java/org/apache/lucene/sandbox/facet/plain/histograms/HistogramCollector.java:
```diff
@@ -53,11 +65,30 @@ public LeafCollector getLeafCollector(LeafReaderContext context) throws IOExcept
+    // We can use multi range traversal logic to collect the histogram on numeric
+    // field indexed as point for MATCH_ALL cases. In future, this can be extended
+    // for Point Range Query cases as well
+    if (query instanceof MatchAllDocsQuery && context.reader().hasDeletions() == false) {
```
Maybe replace this condition with `weight.count(context) == context.reader().maxDoc()` to make it a bit more generic, e.g. it could be a range query that happens to fully match the segment.

jainankitk replied:
Yes we can, but I am anyway planning to address `PointRangeQuery` case more generically as fast followup - https://github.com/apache/lucene/issues/14535. Also with intra-segment search, `leafCollector` can be called multiple times, so I want to ensure that in cases we don't use this optimization has minimal overhead. I am assuming `weight.count(context)` will have some overhead?

jainankitk replied:
Oh yeah, that's a good point! Hadn't thought about queries on other fields. Changed

jainankitk committed:
Improve user-facing docs for geo

Signed-off-by: Ankit Jain <jain.ankitk@gmail.com>

jpountz commented on lucene/sandbox/src/java/org/apache/lucene/sandbox/facet/plain/histograms/PointTreeBulkCollector.java:
```diff
@@ -0,0 +1,227 @@
+
+  static boolean canCollectEfficiently(final PointValues pointValues, final long bucketWidth)
+      throws IOException {
+    // TODO: Do we really need pointValues.getDocCount() == pointValues.size() for this optimization
```
I think so, the collector makes sure to count each doc only once, including docs that have two values that fall into the same bucket.

jainankitk replied:
Ah, thats a good point! I was thinking of cases, when two values are across two buckets, but two values falling in same bucket makes the condition obvious

jpountz commented on lucene/sandbox/src/java/org/apache/lucene/sandbox/facet/plain/histograms/PointTreeBulkCollector.java:
```diff
@@ -0,0 +1,227 @@
+
+    // We want that # leaf nodes is more than # buckets so that we can completely skip over
+    // some of the leaf nodes. Higher this ratio, more efficient it is than naive approach!
+    if ((pointValues.size() / 512) < (leafMaxBucket - leafMinBucket)) {
```
I imagine that 512 is the size of leaves with the default codec. It's a bit annoying to rely on a codec parameter. I'm tempted to replace it with a simpler heuristic like the collector does, something like `leafMaxBucket - leafMinBucket <= 1024`.

jainankitk replied:
> I imagine that 512 is the size of leaves with the default codec. It's a bit annoying to rely on a codec parameter. I'm tempted to replace it with a simpler heuristic like the collector does, something like leafMaxBucket - leafMinBucket <= 1024.

While I share your annoyance of relying on codec parameter, `leafMaxBucket - leafMinBucket <= 1024` works for prior case as it uses dense array that is just function of this. But here, if we collect billion documents,  even ` leafMaxBucket - leafMinBucket  <= 10^5` is fairly reasonable. Maybe we can use something more generic and reasonable like 1000 instead of sticking exactly to the codec parameter?

jpountz replied:
Sounds good to me.

jpountz reviewed (commented):
Interesting idea! I like that you integrated it transparently into the collector, so that users can benefit from it out of the box.

jainankitk committed:
Merge branch 'main' into mrt-collector

jainankitk committed:
Addressing review comments and adding changes entry

Signed-off-by: Ankit Jain <jain.ankitk@gmail.com>

jainankitk commented:
> Interesting idea! I like that you integrated it transparently into the collector, so that users can benefit from it out of the box.

Thanks @jpountz for the review. Had some challenges with the integration initially, but `seWeight` method was pretty useful!

jainankitk committed:
Revert "Improve user-facing docs for geo"

This reverts commit 88d8d34d0c22a5c503448034166acc9958c749f6.

stefanvodita reviewed (commented):
Nice results in the benchmark!

jainankitk commented on lucene/sandbox/src/java/org/apache/lucene/sandbox/facet/plain/histograms/PointTreeBulkCollector.java:
```diff
@@ -0,0 +1,219 @@
+    }
+
+    private void finalizePreviousBucket(byte[] packedValue) {
+      // TODO: Can counter ever be 0?
```
I think this specific TODO, we can leave for now. More like a safety guard. I will try to add unit test in future to validate this, and remove if not needed

Others TODOs were important and have been addressed

jainankitk committed:
Adding @lucene.experimental annotation to PointTreeBulkCollector

Signed-off-by: Ankit Jain <jain.ankitk@gmail.com>

jainankitk commented:
> Oooh those JMH benchy results are nice! Though, it's dangerous testing only on random data -- you can draw random conclusions/results. But it's better than no benchmark! Maybe we should add histogram faceting benchy to Lucene's nightly benchmarks?

Thanks @mikemccand for the feedback. Having histogram faceting benchmark in Lucene's nightly will be great! Created issue https://github.com/mikemccand/luceneutil/issues/375 for following up on this.

jainankitk committed:
Addressing review comments

Signed-off-by: Ankit Jain <jain.ankitk@gmail.com>

jpountz commented on lucene/sandbox/src/java/org/apache/lucene/sandbox/facet/plain/histograms/HistogramCollector.java:
```diff
@@ -53,11 +63,31 @@ public LeafCollector getLeafCollector(LeafReaderContext context) throws IOExcept
+    // We can use multi range traversal logic to collect the histogram on numeric
+    // field indexed as point for MATCH_ALL cases. In future, this can be extended
+    // for Point Range Query cases as well
+    if (context.reader().hasDeletions() == false
```
You don't need this check, Weight#count already accounts for deletions.

jainankitk replied:
Ah, nice!

jpountz commented on lucene/sandbox/src/java/org/apache/lucene/sandbox/facet/plain/histograms/HistogramCollector.java:
```diff
@@ -53,11 +63,31 @@ public LeafCollector getLeafCollector(LeafReaderContext context) throws IOExcept
+    // field indexed as point for MATCH_ALL cases. In future, this can be extended
+    // for Point Range Query cases as well
+    if (context.reader().hasDeletions() == false
+        && weight.count(context) == context.reader().maxDoc()) {
```
I would check weight against `null`, applications may have collectors that don't set a weight.

jainankitk replied:
Added `null` check

jainankitk committed:
Adding null check

Signed-off-by: Ankit Jain <jain.ankitk@gmail.com>

stefanvodita reviewed (approved):
Thank you @jainankitk!

stefanvodita commented on lucene/CHANGES.txt:
```diff
@@ -78,6 +78,8 @@ Optimizations
 ---------------------
 * GITHUB#14418: Quick exit on filter query matching no docs when rewriting knn query. (Pan Guixin)
 
+* GITHUB#14439: Efficient Histogram Collection using multi range traversal over PointTrees
```
You should add your name too!

jainankitk replied:
ah, my bad. Fixed!

jainankitk committed:
Adding name to changelog entry

Signed-off-by: Ankit Jain <jain.ankitk@gmail.com>
